File: mifosng-android/src/main/java/com/mifos/mifosxdroid/dialogfragments/loanchargedialog/LoanChargeDialogFragment.java
/*
 * This project is licensed under the open source MPL V2.
 * See https://github.com/openMF/android-client/blob/master/LICENSE.md
 */

package com.mifos.mifosxdroid.dialogfragments.loanchargedialog;

import android.app.Dialog;
import android.os.Bundle;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.Toast;

import com.mifos.mifosxdroid.R;
import com.mifos.mifosxdroid.core.MifosBaseActivity;
import com.mifos.mifosxdroid.core.ProgressableDialogFragment;
import com.mifos.mifosxdroid.core.util.Toaster;
import com.mifos.mifosxdroid.dialogfragments.chargedialog.OnChargeCreateListener;
import com.mifos.mifosxdroid.uihelpers.MFDatePicker;
import com.mifos.objects.client.ChargeCreationResponse;
import com.mifos.objects.client.Charges;
import com.mifos.services.data.ChargesPayload;
import com.mifos.utils.Constants;
import com.mifos.utils.DateHelper;
import com.mifos.utils.FragmentConstants;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import okhttp3.ResponseBody;

/**
 * Created by nellyk on 1/22/2016.
 * <p/>
 * Use this Dialog Fragment to Create and/or Update charges
 */
public class LoanChargeDialogFragment extends ProgressableDialogFragment implements
        MFDatePicker.OnDatePickListener, LoanChargeDialogMvpView {

    public final String LOG_TAG = getClass().getSimpleName();

    private View rootView;

    @BindView(R.id.chargeNameField)
    Spinner spChargeName;

    @BindView(R.id.amount_due_charge)
    EditText etAmountDue;

    @BindView(R.id.et_date)
    EditText etChargeDueDate;

    @BindView(R.id.et_charge_locale)
    EditText etChargeLocale;

    @BindView(R.id.bt_save_charge)
    Button btnSaveCharge;

    @Inject
    LoanChargeDialogPresenter mLoanChargeDialogPresenter;

    private String dueDateString;

    private DialogFragment mfDatePicker;
    private int chargeId;
    private int loanAccountNumber;
    private HashMap<String, Integer> chargeNameIdHashMap = new HashMap<String, Integer>();
    private String chargeName;
    private Charges createdCharge;
    private List<Integer> dueDateAsIntegerList;

    @Nullable
    private OnChargeCreateListener onChargeCreateListener;

    public static LoanChargeDialogFragment newInstance(int loanAccountNumber) {
        LoanChargeDialogFragment loanChargeDialogFragment = new LoanChargeDialogFragment();
        Bundle args = new Bundle();
        args.putInt(Constants.LOAN_ACCOUNT_NUMBER, loanAccountNumber);
        loanChargeDialogFragment.setArguments(args);
        return loanChargeDialogFragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((MifosBaseActivity) getActivity()).getActivityComponent().inject(this);
        if (getArguments() != null)
            loanAccountNumber = getArguments().getInt(Constants.LOAN_ACCOUNT_NUMBER);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if(getDialog() != null && getDialog().getWindow() != null)
            getDialog().getWindow().setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        if (getArguments() != null)
            loanAccountNumber = getArguments().getInt(Constants.LOAN_ACCOUNT_NUMBER);
        return super.onCreateDialog(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle
            savedInstanceState) {

        // Inflate the layout for this fragment
        if (getActivity().getActionBar() != null)
            getActivity().getActionBar().setDisplayHomeAsUpEnabled(true);
        rootView = inflater.inflate(R.layout.dialog_fragment_charge, null);

        ButterKnife.bind(this, rootView);
        mLoanChargeDialogPresenter.attachView(this);

        inflateDueDate();
        inflateChargesSpinner();

        return rootView;
    }

    @OnClick(R.id.bt_save_charge)
    public void createCharge() {

        if (etAmountDue.getText().toString().isEmpty()) {
            Toaster.show(rootView, getString(R.string.amount)
                    + " " + getString(R.string.error_cannot_be_empty));
            return;
        }
        createdCharge = new Charges();
        createdCharge.setId(chargeId);
        createdCharge.setAmount(Double.parseDouble(etAmountDue.getEditableText().toString()));
        createdCharge.setName(chargeName);

        dueDateAsIntegerList = DateHelper.convertDateAsReverseInteger(dueDateString);
        createdCharge.setDueDate(dueDateAsIntegerList);

        ChargesPayload chargesPayload = new ChargesPayload();
        chargesPayload.setAmount(etAmountDue.getEditableText().toString());
        chargesPayload.setLocale(etChargeLocale.getEditableText().toString());
        chargesPayload.setDueDate(dueDateString);
        chargesPayload.setDateFormat("dd MMMM yyyy");
        chargesPayload.setChargeId(chargeId);

        initiateChargesCreation(chargesPayload);

    }

    @Override
    public void onDatePicked(String date) {
        dueDateString = DateHelper.getDateAsStringUsedForCollectionSheetPayload(date)
                .replace("-", " ");
        etChargeDueDate.setText(dueDateString);
    }

    private void inflateChargesSpinner() {
        mLoanChargeDialogPresenter.loanAllChargesV3(loanAccountNumber);
    }

    private void initiateChargesCreation(ChargesPayload chargesPayload) {
        mLoanChargeDialogPresenter.createLoanCharges(loanAccountNumber, chargesPayload);
    }

    private void inflateDueDate() {
        mfDatePicker = MFDatePicker.newInsance(this);
        String receivedDate = MFDatePicker.getDatePickedAsString();
        dueDateString = DateHelper.getDateAsStringUsedForCollectionSheetPayload(receivedDate)
                .replace("-", " ");
        dueDateAsIntegerList = DateHelper.convertDateAsListOfInteger(dueDateString);
        etChargeDueDate.setText(dueDateString);
    }

    @OnClick(R.id.et_date)
    public void inflateDatePicker() {
        mfDatePicker.show(getActivity().getSupportFragmentManager(), FragmentConstants
                .DFRAG_DATE_PICKER);
    }

    @Override
    public void showAllChargesV3(ResponseBody result) {

        /* Activity is null - Fragment has been detached; no need to do anything. */
        if (getActivity() == null) return;

        Log.d(LOG_TAG, "");

        final List<Charges> charges = new ArrayList<>();
        // you can use this array to populate your spinner
        final ArrayList<String> chargesNames = new ArrayList<String>();
        //Try to get response body
        BufferedReader reader = null;
        StringBuilder sb = new StringBuilder();
        try {
            reader = new BufferedReader(new InputStreamReader(result.byteStream()));
            String line;
            while ((line = reader.readLine()) != null) {
                sb.append(line);
            }
            JSONObject obj = new JSONObject(sb.toString());
            if (obj.has("chargeOptions")) {
                JSONArray chargesTypes = obj.getJSONArray("chargeOptions");
                for (int i = 0; i < chargesTypes.length(); i++) {
                    JSONObject chargesObject = chargesTypes.getJSONObject(i);
                    Charges charge = new Charges();
                    charge.setId(chargesObject.optInt("id"));
                    charge.setName(chargesObject.optString("name"));
                    charges.add(charge);
                    chargesNames.add(chargesObject.optString("name"));
                    chargeNameIdHashMap.put(charge.getName(), charge.getId());
                }

            }
        } catch (Exception e) {
            Log.e(LOG_TAG, "", e);
        }
        final ArrayAdapter<String> chargesAdapter = new ArrayAdapter<String>(getActivity(),
                android.R.layout.simple_spinner_item, chargesNames);
        chargesAdapter.setDropDownViewResource(android.R.layout
                .simple_spinner_dropdown_item);
        spChargeName.setAdapter(chargesAdapter);
        spChargeName.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long
                    l) {
                chargeId = chargeNameIdHashMap.get(chargesNames.get(i));
                chargeName = chargesNames.get(i);
                Log.d("chargesoptionss" + chargesNames.get(i), String.valueOf(chargeId));
                if (chargeId == -1) {
                    Toast.makeText(getActivity(), getString(R.string.error_select_charge)
                            , Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });
    }

    @Override
    public void showLoanChargesCreatedSuccessfully(ChargeCreationResponse chargeCreationResponse) {
        if (onChargeCreateListener != null ) {
            createdCharge.setClientId(chargeCreationResponse.getClientId());
            createdCharge.setId(chargeCreationResponse.getResourceId());
            onChargeCreateListener.onChargeCreatedSuccess(createdCharge);
        } else {
            Toaster.show(rootView, getString(R.string.message_charge_created_success));
        }
        getDialog().dismiss();
    }

    @Override
    public void showError(String s) {
        Toaster.show(rootView, s);
    }

    @Override
    public void showChargeCreatedFailure(String errorMessage) {
        if (onChargeCreateListener != null) {
            onChargeCreateListener.onChargeCreatedFailure(errorMessage);
        } else {
            Toaster.show(rootView, errorMessage);
        }
    }

    @Override
    public void showProgressbar(boolean b) {
        showProgress(b);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mLoanChargeDialogPresenter.detachView();
    }

    public void setOnChargeCreateListener(@Nullable OnChargeCreateListener onChargeCreateListener) {
        this.onChargeCreateListener = onChargeCreateListener;
    }
}

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/dialogfragments/loanchargedialog/LoanChargeDialogFragment.kt
/*
 * This project is licensed under the open source MPL V2.
 * See https://github.com/openMF/android-client/blob/master/LICENSE.md
 */
package com.mifos.mifosxdroid.dialogfragments.loanchargedialog

import android.app.Dialog
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.fragment.app.DialogFragment
import androidx.lifecycle.ViewModelProvider
import com.mifos.core.data.ChargesPayload
import com.mifos.core.objects.client.ChargeCreationResponse
import com.mifos.core.objects.client.Charges
import com.mifos.mifosxdroid.R
import com.mifos.mifosxdroid.core.ProgressableDialogFragment
import com.mifos.mifosxdroid.core.util.Toaster.show
import com.mifos.mifosxdroid.databinding.DialogFragmentChargeBinding
import com.mifos.mifosxdroid.dialogfragments.chargedialog.OnChargeCreateListener
import com.mifos.mifosxdroid.uihelpers.MFDatePicker
import com.mifos.mifosxdroid.uihelpers.MFDatePicker.Companion.newInsance
import com.mifos.mifosxdroid.uihelpers.MFDatePicker.OnDatePickListener
import com.mifos.utils.Constants
import com.mifos.utils.DateHelper
import com.mifos.utils.FragmentConstants
import dagger.hilt.android.AndroidEntryPoint
import okhttp3.ResponseBody
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader

/**
 * Created by nellyk on 1/22/2016.
 *
 *
 * Use this Dialog Fragment to Create and/or Update charges
 */
@AndroidEntryPoint
class LoanChargeDialogFragment : ProgressableDialogFragment(), OnDatePickListener {

    private lateinit var binding: DialogFragmentChargeBinding

    val LOG_TAG = javaClass.simpleName

    private lateinit var viewModel: LoanChargeDialogViewModel

    private lateinit var dueDateString: String
    private var mfDatePicker: DialogFragment? = null
    private var chargeId = 0
    private var loanAccountNumber = 0
    private val chargeNameIdHashMap = HashMap<String?, Int?>()
    private var chargeName: String? = null
    private var createdCharge: Charges? = null
    private var dueDateAsIntegerList: List<Int>? = null
    private var onChargeCreateListener: OnChargeCreateListener? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) loanAccountNumber =
            requireArguments().getInt(Constants.LOAN_ACCOUNT_NUMBER)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        if (arguments != null) loanAccountNumber =
            requireArguments().getInt(Constants.LOAN_ACCOUNT_NUMBER)
        return super.onCreateDialog(savedInstanceState)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {

        // Inflate the layout for this fragment
        if (requireActivity().actionBar != null) requireActivity().actionBar?.setDisplayHomeAsUpEnabled(
            true
        )
        binding = DialogFragmentChargeBinding.inflate(inflater, container, false)
        viewModel = ViewModelProvider(this)[LoanChargeDialogViewModel::class.java]
        inflateDueDate()
        inflateChargesSpinner()

        viewModel.loanChargeDialogUiState.observe(viewLifecycleOwner) {
            when (it) {
                is LoanChargeDialogUiState.ShowAllChargesV3 -> {
                    showProgressbar(false)
                    showAllChargesV3(it.response)
                }

                is LoanChargeDialogUiState.ShowError -> {
                    showProgressbar(false)
                    showError(it.message)
                }

                is LoanChargeDialogUiState.ShowLoanChargesCreatedSuccessfully -> {
                    showProgressbar(false)
                    showLoanChargesCreatedSuccessfully(it.chargeCreationResponse)
                }

                is LoanChargeDialogUiState.ShowProgressbar -> showProgressbar(true)
            }
        }

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.btSaveCharge.setOnClickListener {
            createCharge()
        }

        binding.etDate.setOnClickListener {
            inflateDatePicker()
        }
    }

    private fun createCharge() {
        if (binding.amountDueCharge.text.toString().isEmpty()) {
            show(
                binding.root, getString(R.string.amount)
                        + " " + getString(R.string.error_cannot_be_empty)
            )
            return
        }
        createdCharge = Charges()
        createdCharge?.id = chargeId
        createdCharge?.amount = binding.amountDueCharge.editableText.toString().toDouble()
        createdCharge?.name = chargeName
        dueDateAsIntegerList = DateHelper.convertDateAsReverseInteger(dueDateString)
        createdCharge?.dueDate = dueDateAsIntegerList as List<Int>
        val chargesPayload = ChargesPayload()
        chargesPayload.amount = binding.amountDueCharge.editableText.toString()
        chargesPayload.locale = binding.etChargeLocale.editableText.toString()
        chargesPayload.dueDate = dueDateString
        chargesPayload.dateFormat = "dd MMMM yyyy"
        chargesPayload.chargeId = chargeId
        initiateChargesCreation(chargesPayload)
    }

    override fun onDatePicked(date: String?) {
        dueDateString = DateHelper.getDateAsStringUsedForCollectionSheetPayload(date)
            .replace("-", " ")
        binding.chargeDueDate.text = dueDateString
    }

    private fun inflateChargesSpinner() {
        viewModel.loanAllChargesV3(loanAccountNumber)
    }

    private fun initiateChargesCreation(chargesPayload: ChargesPayload) {
        viewModel.createLoanCharges(loanAccountNumber, chargesPayload)
    }

    private fun inflateDueDate() {
        mfDatePicker = newInsance(this)
        val receivedDate: String? = MFDatePicker.datePickedAsString
        dueDateString = DateHelper.getDateAsStringUsedForCollectionSheetPayload(receivedDate)
            .replace("-", " ")
        dueDateAsIntegerList = DateHelper.convertDateAsListOfInteger(dueDateString)
        binding.chargeDueDate.text = dueDateString
    }


    private fun inflateDatePicker() {
        mfDatePicker?.show(
            requireActivity().supportFragmentManager,
            FragmentConstants.DFRAG_DATE_PICKER
        )
    }

    private fun showAllChargesV3(result: ResponseBody) {

        /* Activity is null - Fragment has been detached; no need to do anything. */
        if (activity == null) return
        Log.d(LOG_TAG, "")
        val charges: MutableList<Charges> = ArrayList()
        // you can use this array to populate your spinner
        val chargesNames = ArrayList<String>()
        //Try to get response body
        var reader: BufferedReader? = null
        val sb = StringBuilder()
        try {
            reader = BufferedReader(InputStreamReader(result.byteStream()))
            var line: String?
            while (reader.readLine().also { line = it } != null) {
                sb.append(line)
            }
            val obj = JSONObject(sb.toString())
            if (obj.has("chargeOptions")) {
                val chargesTypes = obj.getJSONArray("chargeOptions")
                for (i in 0 until chargesTypes.length()) {
                    val chargesObject = chargesTypes.getJSONObject(i)
                    val charge = Charges()
                    charge.id = chargesObject.optInt("id")
                    charge.name = chargesObject.optString("name")
                    charges.add(charge)
                    chargesNames.add(chargesObject.optString("name"))
                    chargeNameIdHashMap[charge.name] = charge.id
                }
            }
        } catch (e: Exception) {
            Log.e(LOG_TAG, "", e)
        }
        val chargesAdapter = ArrayAdapter(
            requireActivity(),
            android.R.layout.simple_spinner_item, chargesNames
        )
        chargesAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        binding.spChargeName.adapter = chargesAdapter
        binding.spChargeName.onItemSelectedListener = object : OnItemSelectedListener {
            override fun onItemSelected(adapterView: AdapterView<*>?, view: View, i: Int, l: Long) {
                chargeId = chargeNameIdHashMap[chargesNames[i]]!!
                chargeName = chargesNames[i]
                Log.d("chargesoptionss" + chargesNames[i], chargeId.toString())
                if (chargeId == -1) {
                    Toast.makeText(
                        activity, getString(R.string.error_select_charge), Toast.LENGTH_SHORT
                    ).show()
                }
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }
    }

    private fun showLoanChargesCreatedSuccessfully(chargeCreationResponse: ChargeCreationResponse) {
        if (onChargeCreateListener != null) {
            createdCharge?.clientId = chargeCreationResponse.clientId
            createdCharge?.id = chargeCreationResponse.resourceId
            createdCharge?.let { onChargeCreateListener?.onChargeCreatedSuccess(it) }
        } else {
            show(binding.root, getString(R.string.message_charge_created_success))
        }
        dialog?.dismiss()
    }

    private fun showError(s: String) {
        show(binding.root, s)
    }

    private fun showChargeCreatedFailure(errorMessage: String) {
        if (onChargeCreateListener != null) {
            onChargeCreateListener?.onChargeCreatedFailure(errorMessage)
        } else {
            show(binding.root, errorMessage)
        }
    }

    private fun showProgressbar(b: Boolean) {
        showProgress(b)
    }

    fun setOnChargeCreateListener(onChargeCreateListener: OnChargeCreateListener?) {
        this.onChargeCreateListener = onChargeCreateListener
    }

    companion object {
        fun newInstance(loanAccountNumber: Int): LoanChargeDialogFragment {
            val loanChargeDialogFragment = LoanChargeDialogFragment()
            val args = Bundle()
            args.putInt(Constants.LOAN_ACCOUNT_NUMBER, loanAccountNumber)
            loanChargeDialogFragment.arguments = args
            return loanChargeDialogFragment
        }
    }
}